File: lib/services/weibo.dart
// 微博相关api的请求
// 由于微博api经过webview请求，所以需要先初始化
import 'dart:async';
import 'dart:convert';
import 'package:webview_flutter/webview_flutter.dart';
import 'package:xany/utils/javascriptChannel.dart';

WebViewController _webViewController;
MyJavascriptChannel _javascriptChannel;

int _id = 100;
Map<int, Completer> _completerMap = {0: Completer()};

int getId() {
  return _id++;
}

void _receivedApiMessage(data) {
  final response = data['response'];
  final id = data['id'];
  final completer = _completerMap[id];
  if (completer == null) return;
  completer.complete(response);
  _completerMap.remove(id);
  print(json.encode(_completerMap));
}

Future _webviewFetch(String url, {options}) {
  final id = getId();
  if (options == null) options = {};
  _webViewController.evaluateJavascript('''
    fetch('$url',${json.encode(options)})
    .then(function(res){ return res.json();})
    .then(function(data){
      ${_javascriptChannel.name}.postMessage(JSON.stringify({
        type:'weiboApi',
        data: { response: data, id: $id}
      }))
    });
  ''');
  _completerMap[id] = Completer();
  return _completerMap[id].future;
}

Future _coovCall(String name, {Map data}) async {
  await _completerMap[0].future; // 等待init成功
  final id = getId();
  if (data == null) data = {};
  _webViewController.evaluateJavascript('''
    window.coov.postData($id, window.coov.$name(${json.encode(data)}));
  ''');
  _completerMap[id] = Completer();
  return _completerMap[id].future;
}

void appendJS(String url) {
  _webViewController.evaluateJavascript('''
  var s = document.createElement('script');
  s.src = '$url';
  document.body.appendChild(s);
  ''');
}

Future initWeiboService(
    WebViewController controller, MyJavascriptChannel javascriptChannel) {
  _webViewController = controller;
  _javascriptChannel = javascriptChannel;
  javascriptChannel.on('weiboApi', _receivedApiMessage);
  appendJS('https://weibo-1252458005.cos.ap-guangzhou.myqcloud.com/weibo.js');
  return _completerMap[0].future;
}

void resetInit() {
  _completerMap[0] = Completer();
}

/// 获取微博
Future getWeibo({Map data}) {
  return _coovCall('getWeibos', data: data);
}

File: lib/services/weibo.js
// 微博接口中转层
window.coov = {
	postData(id, data) {
		if (data instanceof Promise) {
			data.then(d => coovjs.postMessage(JSON.stringify({
				type: 'weiboApi',
				data: { response: d, id: id }
			})))
		} else {
			coovjs.postMessage(JSON.stringify({
				type: 'weiboApi',
				data: { response: data, id: id }
			}))
		}
	},
	fetch(url, options) {
		return fetch(url, options).then(res => res.json());
	},
	objToSearch(obj) {
		if (!obj) return '';
		return Object.keys(obj).map(k => `${k}=${obj[k]}`).join('&')
	},
	handleWeibo(statuses) {
		const div = document.createElement('div')
		const text = statuses.text;
		div.innerHTML = text;
		statuses.text = Array.from(div.childNodes).map(node => {
			const obj = {
				type: '',//emoji|img|video|address|tag|br|text|link|fullText
				url: '',
				content: ''
			}
			switch (node.nodeName) {
				case '#text':
					obj.type = 'text';
					obj.content = node.wholeText;
					break;
				case 'BR':
					obj.type = 'br';
					break;
				case 'SPAN':
					obj.type = 'emoji';
					obj.url = node.children[0].src;
					break;
				case 'A':
					obj.type = 'link';//tag|address|user|img|video|fullText
					if (/^#.+#$/.test(node.innerText)) obj.type = 'tag';
					else if (/^@.+/.test(node.innerText)) obj.type = 'user'
					else if (node.innerText === '全文') obj.type = 'fullText';
					else if (node.childNodes[0].nodeName === 'SPAN') {
						if (node.innerText === '查看图片') obj.type = 'img';
						else if (/^https?:\/\/video\.weibo\.com/.test(node.href)) obj.type = 'video';
						else if (/^http:\/\/weibo\.com\/p\//.test(node.href)) obj.type = 'address'
					}
					obj.content = node.innerText;
					obj.url = node.href;
					break;
				default:
					break;
			}
			return obj;
		});
		if (statuses.retweeted_status) window.coov.handleWeibo(statuses.retweeted_status);
	},
	getWeibos(data) {
		return window.coov.fetch('/feed/friends?' + window.coov.objToSearch(data))
			.then(res => {
				const statuses = res.data.statuses;
				statuses.forEach(window.coov.handleWeibo);
				return res;
			});
	}
}

window.coov.postData(0);
